lib/progressrus.rb

```diff
@@ -2,29 +2,60 @@
 require 'securerandom'
 require 'redis'
 require 'time'
-require_relative "progressrus/store"
 require_relative "progressrus/store/base"
 require_relative "progressrus/store/redis"
 require_relative "progressrus/store/progressbar"
 require_relative "progressrus/core_ext/enumerable"
 
 class Progressrus
+  class InvalidStoreError < StandardError
+    def initialize
+      message = <<~MSG
+        The store needs to implement `persist`, `scope`, `find` and `flush`
+        We have a base class that your store can inherit from:
+          Progressrus::Store::Base
+      MSG
+      super(message)
+    end
+  end
+
   class << self
+    def clear_stores
+      @stores = {}
+    end
+
     def stores
-      @@stores ||= Store.new(Store::Redis.new(::Redis.new(host: ENV["PROGRESSRUS_REDIS_HOST"] || "localhost")))
+      @stores ||= {
+        redis: Store::Redis.new(::Redis.new(host: ENV["PROGRESSRUS_REDIS_HOST"] || "localhost"))
+      }
     end
 
-    def scope(scope, store: :first)
-      stores.find_by_name(store).scope(scope)
+    def add_store(name, store)
+      validate_store!(store)
+      new_stores = stores.merge(name => store)
+      @stores = new_stores
+    end
+
+    def scope(scope, store: :redis)
+      stores[store].scope(scope)
     end
     alias_method :all, :scope
 
-    def find(scope, id, store: :first)
-      stores.find_by_name(store).find(scope, id)
+    def find(scope, id, store: :redis)
+      stores[store].find(scope, id)
+    end
+
+    def flush(scope, id = nil, store: :redis)
+      stores[store].flush(scope, id)
     end
 
-    def flush(scope, id = nil, store: :first)
-      stores.find_by_name(store).flush(scope, id)
+    private
+
+    def validate_store!(store)
```

Comment:
Is it really necessary to validate those?
At least for adapter-like things in the Rails codebase, I haven't seen this pattern of iterating over methods and checking that it implements all required things. This feels a bit against Ruby.

If someone didn't implement `find` on the adapter, you just let them crash next time they call `find` on Progressrus.

Reply:
I'm ok removing it, just wanted to avoid unwanted errors to users of the library.

BTW `ActiveJob` uses this pattern 
https://github.com/rails/rails/blob/382a9563a3bc3dd1982f4db1258e8f0e7f5fac8e/activejob/lib/active_job/queue_adapter.rb#L50-L59